src/pkg/task/execution.go

```diff
@@ -43,3 +53,211 @@ type ExecutionManager interface {
+		Keywords: map[string]interface{}{
+			"ExecutionID": id,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, task := range tasks {
+		if !job.Status(task.Status).Final() {
+			return errors.New(nil).WithCode(errors.PreconditionCode).
+				WithMessage("the execution %d has tasks that aren't in final status, stop the tasks first", id)
+		}
+		if err = e.taskDAO.Delete(ctx, task.ID); err != nil {
+			return err
+		}
+	}
+
+	return e.executionDAO.Delete(ctx, id)
+}
+
+func (e *executionManager) Get(ctx context.Context, id int64) (*Execution, error) {
+	execution, err := e.executionDAO.Get(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return e.populateExecution(ctx, execution), nil
+}
+
+func (e *executionManager) List(ctx context.Context, query *q.Query) ([]*Execution, error) {
+	executions, err := e.executionDAO.List(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	var execs []*Execution
+	for _, execution := range executions {
+		execs = append(execs, e.populateExecution(ctx, execution))
+	}
+	return execs, nil
+}
+
+func (e *executionManager) populateExecution(ctx context.Context, execution *dao.Execution) *Execution {
+	exec := &Execution{
+		ID:            execution.ID,
+		VendorType:    execution.VendorType,
+		VendorID:      execution.VendorID,
+		Status:        execution.Status,
+		StatusMessage: execution.StatusMessage,
+		Metrics:       nil,
+		Trigger:       execution.Trigger,
+		StartTime:     execution.StartTime,
+		EndTime:       execution.EndTime,
+	}
+
+	if len(execution.ExtraAttrs) > 0 {
+		extras := map[string]interface{}{}
+		if err := json.Unmarshal([]byte(execution.ExtraAttrs), &extras); err != nil {
+			log.Errorf("failed to unmarshal the extra attributes of execution %d: %v", execution.ID, err)
+		} else {
+			exec.ExtraAttrs = extras
+		}
+	}
+
+	// if the status isn't null which means the status is set manually, return directly
+	if len(exec.Status) > 0 {
+		return exec
+	}
+
+	// populate task metrics
+	e.populateExecutionMetrics(ctx, exec)
+	// populate status
+	e.populateExecutionStatus(exec)
+	// populate the end time
+	e.populateExecutionEndTime(ctx, exec)
+
+	return exec
+}
+
+func (e *executionManager) populateExecutionMetrics(ctx context.Context, execution *Execution) {
+	scs, err := e.taskDAO.ListStatusCount(ctx, execution.ID)
+	if err != nil {
+		log.Errorf("failed to list status count of execution %d: %v", execution.ID, err)
+		return
+	}
+
+	metrics := &Metrics{}
+	for _, sc := range scs {
+		switch sc.Status {
+		case job.SuccessStatus.String():
+			metrics.SuccessTaskCount = sc.Count
+		case job.ErrorStatus.String():
+			metrics.ErrorTaskCount = sc.Count
+		case job.PendingStatus.String():
+			metrics.PendingTaskCount = sc.Count
+		case job.RunningStatus.String():
+			metrics.RunningTaskCount = sc.Count
+		case job.ScheduledStatus.String():
+			metrics.ScheduledTaskCount = sc.Count
+		case job.StoppedStatus.String():
+			metrics.StoppedTaskCount = sc.Count
+		}
```

Comment:
how about add default clause and return error

Reply:
default clause : just log error?

Reply:
Updated